napari/components/chunk/_loader.py

```diff
@@ -1,52 +1,278 @@
+        We return a ChunkRequest if we performed the load synchronously,
+        otherwise we return None is indicating that an asynchronous load
+        was intitiated. When the async load finishes the layer's
+        on_chunk_loaded() will be called from the GUI thread.
+        """
+        if self._load_synchronously(request):
+            return request
+
+        # Clear any pending requests for this specific data_id.
+        self._clear_pending(request.key.data_id)
+
+        # Sumbit the request asynchronously.
+        self._submit_async(request)
+
+    def _load_synchronously(self, request: ChunkRequest) -> bool:
+        """Return True if we loaded the request synchronously."""
+
+        # If async is enabled, loaded non-ndarray request async.
+        if not self.synchronous and not request.in_memory:
+            return False  # load async
+
+        # Load synchronously.
+        request.load_chunks()
+        return True
+
+    def _submit_async(self, request: ChunkRequest) -> None:
+        """Initiate an asynchronous load of the given request.
+
+        Parameters
+        ----------
+        request : ChunkRequest
+            Contains the arrays to load.
         """
+        LOGGER.debug("ChunkLoader._load_async: %s", request.key)
+
+        # Submit the future, have it call ChunkLoader._done when done.
+        future = self.executor.submit(_chunk_loader_worker, request)
+        future.add_done_callback(self._done)
+
+        # Store the future in case we need to cancel it.
+        self.futures.setdefault(request.key.data_id, []).append(future)
+
+    def _clear_pending(self, data_id: int) -> None:
+        """Clear any pending requests for this data_id.
+
+        Parameters
+        ----------
+        data_id : int
+            Clear all requests associated with this data_id.
+        """
+        LOGGER.debug("ChunkLoader._clear_pending %d", data_id)
+
+        # Get list of futures we submitted to the pool.
+        future_list = self.futures.setdefault(data_id, [])
 
-        request.load_chunks()  # Load all chunks synchronously.
-        return request
+        # Try to cancel all futures in the list, but cancel() will return
+        # False if the task already started running.
+        num_before = len(future_list)
+        future_list[:] = [x for x in future_list if x.cancel()]
+        num_after = len(future_list)
+        num_cleared = num_before - num_after
+
+        # Delete the list entirely if empty
+        if num_after == 0:
+            del self.futures[data_id]
+
+        # Log what we did.
+        if num_before == 0:
+            LOGGER.debug("ChunkLoader.clear_pending: empty")
+        else:
+            LOGGER.debug(
+                "ChunkLoader.clear_pending: %d of %d cleared -> %d remain",
+                num_cleared,
+                num_before,
+                num_after,
+            )
+
+    @staticmethod
+    def _get_request(future: futures.Future) -> Optional[ChunkRequest]:
```

Comment:
It's a staticmethod because it doesn't use self. Pylint will warn if a method doesn't use self. You can either make it a staticmethod or make it a stand-alone function. I usually make it a stand-alone function if it seems like something generic that we might use more than once. And I tend to make it a staticmethod if it seems very tied to class, something we aren't going to use again.

But napari style could offer guidance, when to make a staticmethod (never, always, sometimes?) vs. a stand-alone function. If we wanted to try to be consistent. Either way it will execute identically, just has to live somewhere.

Reply:
In the future I generally would prefer functions, but definitely not a big deal here. As you say — one for the style guide.